syntax = "proto3";

// PreservationPlugin service
service PreservationPlugin {
    rpc Scramble(ToScramble) returns (Scrambled) {}

    rpc DeScramble(Scrambled) returns (NDArray) {}
}


enum ScrambleKind {
    ROW = 0;
    COL = 1;
}


message ToScramble {
    NDArray array = 1;
    ScrambleKind kind = 2;
}



message Scrambled {
    NDArray array = 1;
    Permutation perm = 2;
}



message Permutation {
    bytes data = 1;
    ScrambleKind kind = 2;
}

message NDArray {
    bytes data = 1;
    repeated int32 shape = 2;
}